README.md
## Animate any two icons

mutable_icon is a simplified version of [animate_icons](https://pub.dev/packages/animate_icons "View original Widget"). Watch the demo for more details. It is specifically designed to be included within other widgets that already have their own behavior and only need to animate an icon.

### Demo:

<img src="https://raw.githubusercontent.com/jmouriz/mutable_icon/main/demo/mutable_icon.gif" />


### How to use:

All plugin in your pubspec.yaml 
    
> mutable_icon:

Make the import:

> import 'package:mutable_icon/mutable_icon.dart';

Use the following widget:

```
    MutableIcon(
        startIcon: Icons.add_circle,
        endIcon: Icons.add_circle_outline,
        controller: controller,
        size: 32.0,
        duration: Duration(milliseconds: 400),
        startIconColor: Colors.deepPurple,
        endIconColor: Colors.deepOrange,
        clockwise: false,
        initFrom: InitFrom.end,
    ),
```

# Use MutableIconController
Define MutableIconController to animate b/w start and end icons.

### Define MutableIconController
    
> MutableIconController controller;

### Initialize controller    
    
> controller = MutableIconController();

### Pass controller to widget 
```
MutableIcon(
    startIcon: Icons.add,
    endIcon: Icons.close,
    controller: controller, 
),
```
### Use controller functions
``` 
if (controller.isStart()) {
    controller.animateToEnd();
} else if (controller.isEnd()) {
    controller.animateToStart();
}
```

### Full example
```
import 'package:flutter/material.dart';

import 'package:mutable_icon/mutable_icon.dart';

void main() => runApp(const Example());

class Example extends StatefulWidget {
  const Example({Key? key}) : super(key: key);

  @override
  State<Example> createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  bool _submenu = false;
  late final MutableIconController _controller;

  @override
  void initState() {
    _controller = MutableIconController();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'MutableIcon Example',
      home: Scaffold(
        body: ListView(
          children: [
            ListTile(
              leading: const Icon(
                Icons.wallet,
                color: Colors.black
              ),
              title: const Text('Wallet'),
              trailing: MutableIcon(
                duration: const Duration(milliseconds: 300),
                startIcon: Icons.expand_more,
                endIcon: Icons.expand_less,
                startIconColor: Colors.black,
                endIconColor: Colors.black,
                controller: _controller,
                initFrom: _submenu ? InitFrom.end : InitFrom.start
              ),
              onTap: () {
                setState(() {
                  _submenu = !_submenu;
                  if (_submenu) {
                    _controller.animateToEnd();
                  } else {
                    _controller.animateToStart();
                  }
                });
              },
            ),
            if (_submenu) ListTile(
              contentPadding: const EdgeInsets.only(left: 32),
              leading: const Icon(
                Icons.account_balance,
                color: Colors.black
              ),
              title: const Text('Balance'),
              onTap: () {},
            ),
          ],
        )
      ),
    );
  }
}
```


CHANGELOG.md
## 1.0.0

* First version

## 1.0.1

* Add README content

## 1.1.0

* Add new initFrom method.
* Add example
* Improve API doc
* Update README

## 1.1.1

* Fix format issue.
* Longest package description.


test/mutable_icon_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:mutable_icon/mutable_icon.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/mutable_icon.dart
/// Author: Juan Manuel Mouriz
/// website: tecnologica.com.ar
/// Version: 1.1.1
/// Null-Safety: checked!
/// Prefer Const: checked!

import 'dart:math' as math;

import 'package:flutter/material.dart';

/// Indicates whether to init from the start or from the end.
enum InitFrom {
  /// Init from start to end
  start,

  /// Init from end to start
  end
}

/// Class [MutableIcon]:
///
/// [startIcon]: The IconData that will be visible before animation starts.
///
/// [endIcon]: The IconData that will be visible after animation ends.
///
/// [controller]: MutableIcon controller.
///
/// [startIconColor]: The color to be used for the [startIcon].
///
/// [endIconColor]: The color to be used for the [endIcon].
///
/// [duration]: The duration for which the animation runs.
///
/// [size]: The size of the icon that are to be shown.
///
/// [clockwise]: If the animation runs in the clockwise or anticlockwise
/// direction.
///
/// [initFrom] Used to set initial state. Indicates whether to init from the
/// start or from the end. Default is [initFrom.start].
class MutableIcon extends StatefulWidget {
  /// Creates a [MutableIcon] widget.
  ///
  /// [startIcon]: The IconData that will be visible before animation starts.
  ///
  /// [endIcon]: The IconData that will be visible after animation ends.
  ///
  /// [controller]: MutableIcon controller.
  ///
  /// [startIconColor]: The color to be used for the [startIcon].
  ///
  /// [endIconColor]: The color to be used for the [endIcon].
  ///
  /// [duration]: The duration for which the animation runs.
  ///
  /// [size]: The size of the icon that are to be shown.
  ///
  /// [clockwise]: If the animation runs in the clockwise or anticlockwise
  /// direction.
  ///
  /// [initFrom]: Used to set initial state. Indicates whether to init from the
  /// start or from the end. Default is [initFrom.start].
  const MutableIcon({
    Key? key,

    /// The IconData that will be visible before animation starts
    required this.startIcon,

    /// The IconData that will be visible after animation ends
    required this.endIcon,

    /// MutableIcon controller
    required this.controller,

    /// The color to be used for the [startIcon]
    this.startIconColor,

    /// The color to be used for the [endIcon]
    this.endIconColor,

    /// The duration for which the animation runs
    this.duration = const Duration(milliseconds: 300),

    /// The size of the icon that are to be shown
    this.size = 24.0,

    /// If the animation runs in the clockwise or anticlockwise direction
    this.clockwise = false,

    /// Used to set initial state. Indicates whether to init from the start
    /// or from the end. Default is [initFrom.start]
    this.initFrom = InitFrom.start,
  }) : super(key: key);

  final IconData startIcon, endIcon;
  final MutableIconController controller;
  final Color? startIconColor, endIconColor;
  final Duration duration;
  final double size;
  final bool clockwise;
  final InitFrom initFrom;

  @override
  _MutableIconState createState() => _MutableIconState();
}

class _MutableIconState extends State<MutableIcon>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void initState() {
    _controller = AnimationController(
      vsync: this,
      duration: widget.duration,
      value: widget.initFrom == InitFrom.start ? 0.0 : 1.0,
      lowerBound: 0.0,
      upperBound: 1.0,
    );
    _controller.addListener(() {
      if (mounted) {
        setState(() {});
      }
    });
    initControllerFunctions();
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  initControllerFunctions() {
    widget.controller.animateToEnd = () {
      if (mounted) {
        _controller.forward();
        return true;
      } else {
        return false;
      }
    };
    widget.controller.animateToStart = () {
      if (mounted) {
        _controller.reverse();
        return true;
      } else {
        return false;
      }
    };
    widget.controller.isStart = () => _controller.value == 0.0;
    widget.controller.isEnd = () => _controller.value == 1.0;
  }

  @override
  Widget build(BuildContext context) {
    double start = _controller.value;
    double end = 1.0 - _controller.value;
    double startOpacity = start;
    double endOpacity = end;
    double startAngle = math.pi / 180 * (180 * startOpacity);
    double endAngle = math.pi / 180 * (180 * endOpacity);

    Widget first() {
      final icon = Icon(
        widget.startIcon,
        size: widget.size,
        color: widget.startIconColor ?? Theme.of(context).primaryColor,
      );
      return Transform.rotate(
          angle: widget.clockwise ? startAngle : -startAngle,
          child: Opacity(
            opacity: endOpacity,
            child: icon,
          ));
    }

    Widget second() {
      final icon = Icon(
        widget.endIcon,
        size: widget.size,
        color: widget.endIconColor ?? Theme.of(context).primaryColor,
      );
      return Transform.rotate(
        angle: widget.clockwise ? -endAngle : endAngle,
        child: Opacity(opacity: startOpacity, child: icon),
      );
    }

    return Stack(
      alignment: Alignment.center,
      children: [
        start == 1 && end == 0 ? second() : first(),
        start == 0 && end == 1 ? first() : second(),
      ],
    );
  }
}

class MutableIconController {
  late bool Function() animateToStart = () => false;
  late bool Function() animateToEnd = () => false;
  late bool Function() isStart, isEnd;
}


example/main.dart
import 'package:flutter/material.dart';

import 'package:mutable_icon/mutable_icon.dart';

void main() => runApp(const Example());

class Example extends StatefulWidget {
  const Example({Key? key}) : super(key: key);

  @override
  State<Example> createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  bool _submenu = false;
  late final MutableIconController _controller;

  @override
  void initState() {
    _controller = MutableIconController();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'MutableIcon Example',
      home: Scaffold(
          body: ListView(
        children: [
          ListTile(
            leading: const Icon(Icons.wallet, color: Colors.black),
            title: const Text('Wallet'),
            trailing: MutableIcon(
                duration: const Duration(milliseconds: 300),
                startIcon: Icons.expand_more,
                endIcon: Icons.expand_less,
                startIconColor: Colors.black,
                endIconColor: Colors.black,
                controller: _controller,
                initFrom: _submenu ? InitFrom.end : InitFrom.start),
            onTap: () {
              setState(() {
                _submenu = !_submenu;
                if (_submenu) {
                  _controller.animateToEnd();
                } else {
                  _controller.animateToStart();
                }
              });
            },
          ),
          if (_submenu)
            ListTile(
              contentPadding: const EdgeInsets.only(left: 32),
              leading: const Icon(Icons.account_balance, color: Colors.black),
              title: const Text('Balance'),
              onTap: () {},
            ),
        ],
      )),
    );
  }
}
